File: mandel.c
#include <stdio.h>

void printdensity(double d) {
  if (d > 127) {
    putchar(32);
  } else if (d > 8) {
    putchar(46);
  } else if (d > 4) {
    putchar(43);
  } else {
    putchar(42);
  }
}

void newline() {
    putchar(10);
}

double mandelconverge(double creal, double cimag) {
    double r = creal;
    double i = cimag;
    int iter = 0;
    for (; iter < 255; iter++) {
        if (r*r + i*i > 4) {
            return iter;
        }
        double next_r = r*r - i*i + creal;
        i = 2*r*i + cimag;
        r = next_r;
    }
    return iter;
}

void mandel(double xmin, double ymin, double xstep, double ystep, int xsteps, int ysteps) {
    double xmax = xmin + xstep * xsteps;
    double ymax = ymin + ystep * ysteps;
    for (int iy = 0; iy < ysteps; iy++) {
        double y = iy * (ymax - ymin) * ystep + ymin;
        for (int ix = 0; ix < xsteps; ix++) {
            double x = ix * (xmax - xmin) * xstep + xmin;
            printdensity(mandelconverge(x,y));
        }
        newline();
    }
}

void main() {
    // mandel(-2.3, -2.0, 0.025, 0.05, 78, 40);
    mandel(-2.3, -2.0, 0.025 / 2, 0.05 / 2, 156, 80);
}

File: mandel.rs
fn printdensity(d: i32) {
    if d > 127 {
      print!("{}", ' ');
    } else if d > 8 {
      print!("{}", '.');
    } else if d > 4 {
      print!("{}", '+');
    } else {
      print!("{}", '*');
    }
}
  
fn newline() {
    println!("");
}
  
fn mandelconverge(creal: f64, cimag: f64) -> i32 {
    let mut r = creal;
    let mut i = cimag;
    const MAX_ITER: i32 = 255;
    for iter in 0..MAX_ITER {
        if r*r + i*i > 4. {
            return iter;
        }
        let next_r = r*r - i*i + creal;
        i = 2.*r*i + cimag;
        r = next_r;
    }
    MAX_ITER
}
  
fn mandel(xmin: f64, ymin: f64, xstep: f64, ystep: f64, xsteps: i32, ysteps: i32) {
    let xmax = xmin + xstep * xsteps as f64;
    let ymax = ymin + ystep * ysteps as f64;
    for iy in 0..ysteps {
        let y = iy as f64 * (ymax - ymin) * ystep as f64 + ymin;
        for ix in 0..xsteps {
            let x = ix as f64 * (xmax - xmin) * xstep as f64 + xmin;
            printdensity(mandelconverge(x,y));
        }
        newline();
    }
}
  
fn main() {
    mandel(-2.3, -2.0, 0.025 / 2., 0.05 / 2., 156, 80);
}